### PR TITLE
Fix #258 by adding 3 measures against DoS attacks

### LINKED ISSUE #258: Server memory leak when using AWS Fargate
Hi, the server seems to have a memory leak, we can see the memory rise over an hour before crashing and starting again. The final crash complains about `Too many open files`

The server had no application running on it during this time.

Another thing to note is the constant high CPU usage.

The server is running in an AWS Fargate container with 2CPU's and 4096Mb RAM

CPU/Memory usage 24h
image

Crash logs
```
E0303 10:35:57.812868828      25 tcp_server_posix.cc:213]    Failed accept4: Too many open files
--
2022-03-03 10:38:27,069 - FederatedServer - INFO - starting insecure server at 0.0.0.0:8002
deployed FL server trainer.
2022-03-03 10:38:27,075 - FedAdminServer - INFO - Starting Admin Server 0.0.0.0 on Port 8003
2022-03-03 10:38:27,075 - root - INFO - Server started
/app/start_service.sh: line 10:     8 Killed                  python -u -m nvflare.private.fed.app.server.server_train -m /workspace -s fed_server.json --set secure_train=false config_folder=config host=0.0.0.0
2022-03-03 10:43:29,115 - ClientManager - INFO - Re-activate dead client:ip-10-0-246-19.ap-southeast-2.compute.internal with token: 8de2503e-cb43-4cfd-91a9-14f0b4ae98be
2022-03-03 10:43:52,231 - ClientManager - INFO - Re-activate dead client:ip-10-0-128-176.ap-southeast-2.compute.internal with token: 77d51ef9-3ac3-4026-b26f-907eaa7cbc04
E0303 11:34:48.528427869      25 tcp_server_posix.cc:213]    Failed accept4: Too many open files
2022-03-03 11:37:56,203 - FederatedServer - INFO - starting insecure server at 0.0.0.0:8002
deployed FL server trainer.
2022-03-03 11:37:56,209 - FedAdminServer - INFO - Starting Admin Server 0.0.0.0 on Port 8003
2022-03-03 11:37:56,209 - root - INFO - Server started
/app/start_service.sh: line 10:     8 Killed                  python -u -m nvflare.private.fed.app.server.server_train -m /workspace -s fed_server.json --set secure_train=false config_folder=config host=0.0.0.0
2022-03-03 11:39:37,262 - ClientManager - INFO - Re-activate dead client:ip-10-0-128-176.ap-southeast-2.compute.internal with token: 77d51ef9-3ac3-4026-b26f-907eaa7cbc04
2022-03-03 11:39:40,542 - ClientManager - INFO - Re-activate dead client:ip-10-0-246-19.ap-southeast-2.compute.internal with token: 8de2503e-cb43-4cfd-91a9-14f0b4ae98be
E0303 12:34:14.240603486      24 tcp_server_posix.cc:213]    Failed accept4: Too many open files
2022-03-03 12:37:04,335 - FederatedServer - INFO - starting insecure server at 0.0.0.0:8002
deployed FL server trainer.
2022-03-03 12:37:04,341 - FedAdminServer - INFO - Starting Admin Server 0.0.0.0 on Port 8003
2022-03-03 12:37:04,368 - root - INFO - Server started
2022-03-03 12:42:02,992 - ClientManager - INFO - Re-activate dead client:ip-10-0-128-176.ap-southeast-2.compute.internal with token: 77d51ef9-3ac3-4026-b26f-907eaa7cbc04
2022-03-03 12:42:05,714 - ClientManager - INFO - Re-activate dead client:ip-10-0-246-19.ap-southeast-2.compute.internal with token: 8de2503e-cb43-4cfd-91a9-14f0b4ae98be
E0303 13:33:17.078870193      24 tcp_server_posix.cc:213]    Failed accept4: Too many open files
2022-03-03 13:36:20,660 - FederatedServer - INFO - starting insecure server at 0.0.0.0:8002
deployed FL server trainer.
2022-03-03 13:36:20,666 - FedAdminServer - INFO - Starting Admin Server 0.0.0.0 on Port 8003
2022-03-03 13:36:20,666 - root - INFO - Server started
2022-03-03 13:41:27,090 - ClientManager - INFO - Re-activate dead client:ip-10-0-128-176.ap-southeast-2.compute.internal with token: 77d51ef9-3ac3-4026-b26f-907eaa7cbc04
2022-03-03 13:41:34,427 - ClientManager - INFO - Re-activate dead client:ip-10-0-246-19.ap-southeast-2.compute.internal with token: 8de2503e-cb43-4cfd-91a9-14f0b4ae98be
E0303 14:32:29.195437902      27 tcp_server_posix.cc:213]    Failed accept4: Too many open files
2022-03-03 14:35:29,825 - FederatedServer - INFO - starting insecure server at 0.0.0.0:8002
deployed FL server trainer.
2022-03-03 14:35:29,831 - FedAdminServer - INFO - Starting Admin Server 0.0.0.0 on Port 8003
2022-03-03 14:35:29,831 - root - INFO - Server started
/app/start_service.sh: line 10:     8 Killed                  python -u -m nvflare.private.fed.app.server.server_train -m /workspace -s fed_server.json --set secure_train=false config_folder=config host=0.0.0.0
2022-03-03 14:37:50,369 - ClientManager - INFO - Re-activate dead client:ip-10-0-128-176.ap-southeast-2.compute.internal with token: 77d51ef9-3ac3-4026-b26f-907eaa7cbc04
2022-03-03 14:37:58,754 - ClientManager - INFO - Re-activate dead client:ip-10-0-246-19.ap-southeast-2.compute.internal with token: 8de2503e-cb43-4cfd-91a9-14f0b4ae98be
E0303 15:31:39.658074612      26 tcp_server_posix.cc:213]    Failed accept4: Too many open files
2022-03-03 15:34:46,429 - FederatedServer - INFO - starting insecure server at 0.0.0.0:8002
deployed FL server trainer.
2022-03-03 15:34:46,435 - FedAdminServer - INFO - Starting Admin Server 0.0.0.0 on Port 8003
2022-03-03 15:34:46,435 - root - INFO - Server started
/app/start_service.sh: line 10:     8 Killed                  python -u -m nvflare.private.fed.app.server.server_train -m /workspace -s fed_server.json --set secure_train=false config_folder=config host=0.0.0.0
2022-03-03 15:38:24,369 - ClientManager - INFO - Re-activate dead client:ip-10-0-246-19.ap-southeast-2.compute.internal with token: 8de2503e-cb43-4cfd-91a9-14f0b4ae98be
2022-03-03 15:39:12,769 - ClientManager - INFO - Re-activate dead client:ip-10-0-128-176.ap-southeast-2.compute.internal with token: 77d51ef9-3ac3-4026-b26f-907eaa7cbc04
E0303 16:30:55.183641326      26 tcp_server_posix.cc:213]    Failed accept4: Too many open files
2022-03-03 16:34:01,946 - FederatedServer - INFO - starting insecure server at 0.0.0.0:8002
deployed FL server trainer.
2022-03-03 16:34:01,952 - FedAdminServer - INFO - Starting Admin Server 0.0.0.0 on Port 8003
2022-03-03 16:34:01,952 - root - INFO - Server started
```

### PR DESCRIPTION
**Pull request by nvidianz**
1. Message size is limited to 0.5 GB
2. Connection is timed out after idling for 10 seconds
3. Number of admin connections is limited to 128

### CODE REVIEW

**Comment by nvidianz** [comment]
Obsolete. Replaced by https://github.com/NVIDIA/NVFlare/pull/295

**Review by yanchengnv — APPROVED** [review]

File: nvflare/fuel/hci/conn.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import logging
```
Comment (YuanTingHsieh) [inline nvflare/fuel/hci/conn.py:15]: this line is not needed, please remove

File: nvflare/fuel/hci/conn.py [hunk 2]
```diff
@@ -27,16 +28,20 @@
 
 LINE_END = "\x03"  # Indicates the end of the line (end of text)
 ALL_END = "\x04"  # Marks the end of a complete transmission (End of Transmission)
-
-
 MAX_MSG_SIZE = 1024
+MAX_DATA_SIZE = 512*1024*1024
```
Comment (YuanTingHsieh) [inline nvflare/fuel/hci/conn.py:32]: As pointed out by the unit test (https://github.com/NVIDIA/NVFlare/runs/5517455146?check_suite_focus=true), it suggest changing this line to

`512 * 1024 * 1024`

File: nvflare/fuel/hci/server/hci.py [hunk 3]
```diff
@@ -68,8 +83,13 @@ def handle(self):
 
             if not conn.ended:
                 conn.close()
-        except BaseException:
-            traceback.print_exc()
+        except BaseException as exc:
+            self.logger.error(f"Admin connection terminated due to exception: {str(exc)}")
+            if self.logger.getEffectiveLevel() <= logging.DEBUG:
+                self.logger.exception("Admin connection error")
```
Comment (YuanTingHsieh) [inline nvflare/fuel/hci/server/hci.py:89]: why do we need these two lines?

All the levels below error would already have the "self.logger.error" printed out.

**Review by YuanTingHsieh — CHANGES_REQUESTED** [review]
Thanks for the fix, mostly looks good to me, just need minor improvements.

**Review by nvkevlu — APPROVED** [review]

File: nvflare/fuel/hci/server/hci.py (hunk 3 above)
Comment (nvidianz) [reply nvflare/fuel/hci/server/hci.py:89]: The purpose is to only print traceback in debug level so it doesn't clog up prod logs.

**Review by YuanTingHsieh — APPROVED** [review]

**Review by SYangster — APPROVED** [review]

**Pull request merged**